### PR TITLE
[SPARK-18403][SQL] Fix unsafe data false sharing issue in ObjectHashAggregateExec

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

This PR fixes a random OOM issue occurred while running `ObjectHashAggregateSuite`.

This issue can be steadily reproduced under the following conditions:

1. The aggregation must be evaluated using `ObjectHashAggregateExec`;
2. There must be an input column whose data type involves `ArrayType` (an input column of `MapType` may even cause SIGSEGV);
3. Sort-based aggregation fallback must be triggered during evaluation.

The root cause is that while falling back to sort-based aggregation, we must sort and feed already evaluated partial aggregation buffers living in the hash map to the sort-based aggregator using an external sorter. However, the underlying mutable byte buffer of `UnsafeRow`s produced by the iterator of the external sorter is reused and may get overwritten when the iterator steps forward. After the last entry is consumed, the byte buffer points to a block of uninitialized memory filled by `5a`. Therefore, while reading an `UnsafeArrayData` out of the `UnsafeRow`, `5a5a5a5a` is treated as array size and triggers a memory allocation for a ridiculously large array and immediately blows up the JVM with an OOM.

To fix this issue, we only need to add `.copy()` accordingly.

## How was this patch tested?

New regression test case added in `ObjectHashAggregateSuite`.

### CODE REVIEW

**Comment by liancheng** [comment]
The last build failure was caused by a logical conflict with #15703. We don't really have any aggregate functions that don't support partial aggregation now after merging #15703, while the re-enabled test cases still checks for that condition.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ObjectHashAggregateSuite.scala [hunk 1]
```diff
@@ -325,70 +320,67 @@ class ObjectHashAggregateSuite
...
+                // Enables sort-based aggregation fallback to obtain another result to be checked.
+                withSQLConf(SQLConf.OBJECT_AGG_SORT_BASED_FALLBACK_THRESHOLD.key -> "3") {
+                  // Here we are not reusing `aggDf` because the physical plan in `aggDf` is
+                  // cached and won't be re-planned using the new fallback threshold.
+                  actual2 = doAggregation(df).collect().toSeq
+                }
               }
+
+              doubleSafeCheckRows(actual1, expected, 1e-4)
+              doubleSafeCheckRows(actual2, expected, 1e-4)
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ObjectHashAggregateSuite.scala]: All the changes made above in this file are used to resolve a logical conflict with PR #15703. We don't really have any aggregate functions that don't support partial aggregation now after merging #15703, must update the tests to reflect that.

**Comment by liancheng** [comment]
cc @yhuai @cloud-fan

**Comment by dongjoon-hyun** [comment]
Thank you for fixing this, @liancheng !

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
The last build failure was caused by YARN tests.

**Comment by liancheng** [comment]
Also cc @davies and @sameeragarwal.

**Comment by liancheng** [comment]
A similar alternative fix @yhuai proposed is to convert the underlying `UnsafeRow` into a safe row (i.e. `GenericInternalRow` in this case) using a projection instead of simply adding a `.copy()`. In this way, we prevent adding unsafe data into a safe row, which is in general safer. This approach further affects performance, though.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala [hunk 2]
```diff
@@ -262,7 +262,9 @@ class SortBasedAggregator(
           // Firstly, update the aggregation buffer with input rows.
           while (hasNextInput &&
             groupingKeyOrdering.compare(inputIterator.getKey, groupingKey) == 0) {
-            processRow(result.aggregationBuffer, inputIterator.getValue)
+            // Since `inputIterator.getValue` is an `UnsafeRow` whose underlying buffer will be
+            // overwritten when `inputIterator` steps forward, we need to do a deep copy here.
+            processRow(result.aggregationBuffer, inputIterator.getValue.copy())
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala]: So the problem is, during `processRow` we cache the input row somehow?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala]: I think it's caused by `MutableProjection`? As `MutableProjection` may keep an "pointer" that points to a memory region of an unsafe row. Maybe we can fix this bug by https://github.com/apache/spark/pull/15082?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala]: nvm, #15082 needs some significant refactor, we should get this fix in 2.1 first.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ObjectHashAggregateSuite.scala [hunk 3]
```diff
@@ -425,7 +417,35 @@ class ObjectHashAggregateSuite
...
+    // SPARK-18403: An unsafe data false sharing issue may trigger OOM / SIGSEGV when evaluating
+    // certain aggregate functions. To reproduce this issue, the following conditions must be
+    // met:
+    //
+    //  1. The aggregation must be evaluated using `ObjectHashAggregateExec`;
+    //  2. There must be an input column whose data type involves `ArrayType` or `MapType`;
+    //  3. Sort-based aggregation fallback must be triggered during evaluation.
+    withSQLConf(
+      SQLConf.USE_OBJECT_HASH_AGG.key -> "true",
+      SQLConf.OBJECT_AGG_SORT_BASED_FALLBACK_THRESHOLD.key -> "1"
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ObjectHashAggregateSuite.scala]: not related to this PR, but the config name looks weird, how about `OBJECT_AGG_FALLBACK_TO_SORT_THRESHOLD`

**Comment by cloud-fan** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by liancheng** [comment]
@cloud-fan @dongjoon-hyun Thanks for the review!

**Pull request closed without merging**